Ensure psutil for the  process is accurate

echarles opened the pull request:
Fixes https://github.com/ipython/ipykernel/issues/912

This add to the control_thread a map of processes to allow the requesting the cpu_usage in a way psutil excepts (the cpu is only correct if is requested on the same process object).

mlucool commented on ipykernel/kernelbase.py:
```diff
@@ -76,6 +76,8 @@ class Kernel(SingletonConfigurable):
     # attribute to override with a GUI
     eventloop = Any(None)
 
+    processes = {}
```
Why is this map needed?

mlucool replied:
Maybe not a big deal, but this seems a bit "append" only (i.e. a process object is never removed)

echarles replied:
> Maybe not a big deal, but this seems a bit "append" only (i.e. a process object is never removed)

https://github.com/ipython/ipykernel/pull/937/commits/17358fb26e2a362691c3f59e4aa0615d6abaf50c addresses this.

echarles commented:
> self.processes = {process.pid: self.processes.get(process.pid, process) for process in all_processes}

I have used that block code and the CPU does not return the correct value. 

...or should I also remove https://github.com/ipython/ipykernel/pull/937/files#diff-874f94d86eaa767fb34865311dd8fd8c0d9a2fa6c49bc21f28690058f3dfa83cR961 as you said ? (I am not sure to get the correct link, it points to just one line).

mlucool commented:
This works for me:
```
import subprocess
from time import sleep

import psutil


class Foo:
    processes = {}

    def update(self):
        current_process = psutil.Process()
        all_processes = [current_process] + current_process.children(recursive=True)
        self.processes = {
            process.pid: self.processes.get(process.pid, process)
            for process in all_processes
        }

    def print_cpu(self):
        print("-----")
        self.update()
        for p, v in self.processes.items():
            print(f"{p}: {v.cpu_percent()}")


def work():
    y = 0
    for x in range(10):
        y += x

    return y


f = Foo()
f.print_cpu()
work()
f.print_cpu()
# Let's use CPU
p = subprocess.Popen(
    ["dd", "if=/dev/zero", "of=/dev/null"],
    stdout=subprocess.PIPE,
    stderr=subprocess.STDOUT,
)
f.print_cpu()
sleep(0.1)
work()
f.print_cpu()
p.kill()
p.wait()
work()
sleep(2)
f.print_cpu()  # Less work
```

Output:
```
-----
24567: 0.0
-----
24567: 87.9
-----
24567: 65.0
127672: 0.0
-----
24567: 23.8
127672: 95.1
-----
24567: 1.5
```

Basically, if we are going to prune on every run, I think we should prune + keep the list up to date to limit the number of times we need to iterate these datastructures.

echarles commented:
@mlucool I have retested and now it work as expected on my env. Thx for the suggestion and better implementation. I have committed those changes.

Anything else before asking a review?

mlucool commented on ipykernel/kernelbase.py:
```diff
@@ -953,11 +955,15 @@ async def debug_request(self, stream, ident, parent):
         try:
             # psutil.Process methods will either return...
-            metric_value = getattr(process, name)()
+            pid = process.pid
+            p = self.processes.get(pid, None)
+            if not p:
+                self.processes[pid] = process
+                p = self.processes.get(pid)
```
These lines are not need as we ensure the a single, reused process below

echarles replied:
Fixed. Thx.

echarles commented on ipykernel/kernelbase.py:
```diff
@@ -971,18 +977,23 @@ async def usage_request(self, stream, ident, parent):
-        process_metric_value = self.get_process_metric_value
+        # Ensure 1) self.processes is updated to only current subprocesses
+        # and 2) we reuse processes when possible (needed for accurate CPU)
+        self.processes = {process.pid: self.processes.get(process.pid, process) for process in all_processes}
```
Removed. Thx.

echarles commented on ipykernel/kernelbase.py:
```diff
@@ -971,18 +977,23 @@ async def usage_request(self, stream, ident, parent):
+        }
         reply_content["kernel_cpu"] = sum(
-            [process_metric_value(process, "cpu_percent", None) for process in all_processes]
+            [self.get_process_metric_value(process, "cpu_percent", None) for process in all_processes]
```
Fixed. Thx.

blink1073 commented:
Docs failure is unrelated, but the pre-commit ones are.

echarles commented:
I have fixed 2 format issues and added a type to the new `processes` class field. pre-commit is now happy, but the typed class field generates other errors, eg on https://github.com/ipython/ipykernel/runs/6805383473?check_suite_focus=true

```
    processes: dict[str, psutil.Process] = {}
E   TypeError: 'type' object is not subscriptable
```

blink1073 commented:
Yeah you need to use `typings.Dict` before Python 3.8.

echarles commented:
@blink1073 CI is now green.

blink1073 reviewed (approved):
Thanks!

blink1073 commented:
I'll cut a new release early next week.

echarles commented:
@blink1073 👍 

> I'll cut a new release early next week.

Super!
